ethereum/contracts/relayer/coreRelayer/CoreRelayer.sol

```diff
@@ -1,68 +1,41 @@
- *         Delivery ---------------^
- */
-
-contract CoreRelayer is CoreRelayerSendOverloads, CoreRelayerDelivery, ERC1967Upgrade {
-    error ImplementationAlreadyInitialized();
-
-    constructor(address _forwardWrapper) CoreRelayerGetters(_forwardWrapper) {}
-
-    // this function needs to be exposed for an upgrade to pass
-    function initialize() public virtual initializer {}
-
-    modifier initializer() {
-        address impl = ERC1967Upgrade._getImplementation();
-
-        if (isInitialized(impl)) {
-            revert ImplementationAlreadyInitialized();
-        }
-
-        setInitialized(impl);
-
-        _;
-    }
-
-    function submitContractUpgrade(bytes memory vaa) public {
-        (bool success, bytes memory reason) = getWormholeRelayerCallerAddress().delegatecall(
-            abi.encodeWithSignature("submitContractUpgrade(bytes)", vaa)
-        );
-        require(success, string(reason));
-    }
-
-    function registerCoreRelayerContract(bytes memory vaa) public {
-        (bool success, bytes memory reason) = getWormholeRelayerCallerAddress().delegatecall(
-            abi.encodeWithSignature("registerCoreRelayerContract(bytes)", vaa)
-        );
-        require(success, string(reason));
-    }
-
-    function setDefaultRelayProvider(bytes memory vaa) public {
-        (bool success, bytes memory reason) = getWormholeRelayerCallerAddress().delegatecall(
-            abi.encodeWithSignature("setDefaultRelayProvider(bytes)", vaa)
-        );
-        require(success, string(reason));
-    }
-
-    function submitRecoverChainId(bytes memory vaa) public {
-        (bool success, bytes memory reason) = getWormholeRelayerCallerAddress().delegatecall(
-            abi.encodeWithSignature("submitRecoverChainId(bytes)", vaa)
-        );
-        require(success, string(reason));
-    }
+pragma solidity ^0.8.19;
+
+import {IWormholeRelayer} from "../../interfaces/relayer/IWormholeRelayer.sol";
+
+import {getDefaultRelayProviderState} from "./CoreRelayerStorage.sol";
+import {CoreRelayerGovernance} from "./CoreRelayerGovernance.sol";
+import {CoreRelayerSend} from "./CoreRelayerSend.sol";
+import {CoreRelayerDelivery} from "./CoreRelayerDelivery.sol";
+import {CoreRelayerBase} from "./CoreRelayerBase.sol";
+
+//CoreRelayerGovernance inherits from ERC1967Upgrade, i.e. this is a proxy contract!
+contract CoreRelayer is
+  CoreRelayerGovernance,
+  CoreRelayerSend,
+  CoreRelayerDelivery,
+  IWormholeRelayer
+{
+  address private immutable initialDefaultRelayProvider;
+
+  //the only normal storage variable - everything else uses slot pattern
+  //no point doing it for this one since it is entirely one-off and of no interest to the rest
+  //  of the contract and it also can't accidentally be moved because we are at the bottom of
+  //  the inheritance hierarchy here
+  bool private initialized;
+
+  constructor(
+    address wormhole,
+    address _initialDefaultRelayProvider
```

Comment:
I'm a bit out of date on decisions related to CREATE2 deployment, but this means that we introduce a dependency on the address of the relay provider to get a deterministic address for the `CoreRelayer` proxy, right? Oh, and the wormhole core address is a problem too now that I think about it. I think we should move these to the `initialize` parameters.

Reply:
That's true, but do we care about the address of the logic contracts? The proxies should all be deployed to the same, but logic contract addresses shouldn't really matter, no?

Reply:
Yeah, the proxies addresses should be fine I think. I think at some point this would have posed a problem. I have to get up to speed on the current deployment logic. If this still poses a problem, we should be able to solve it by adjusting the way the proxy is deployed, I think.